get_balance - serialize response & deserialize request

howjmay opened the pull request:
#870 

json/get_balances.c/
    add `json_get_balances_deserialize_request()`
    add `json_get_balances_serialize_response()`

helper.c/ 
    add `utarray_uint64_to_json_array()`

howjmay committed:
implement serializaztion and deserialization funcs

implement funcs of serializing response and deserializing request
functions, but the unit test hasn't been implemented yet.

implemet json_get_size_t_num() to get size_t obj from cJSON obj

howjmay committed:
Add get_balances unit test and needed helper func

In order to convert uint64_t utarray to json array, I
implemented utarray_uint64_to_json_array() to take this job.

howjmay committed:
Merge branch 'develop' into get_balance_serialize_res_deserialize_req

oopsmonk commented on cclient/serialization/json/get_balances.c:
```diff
@@ -56,6 +56,95 @@ retcode_t json_get_balances_serialize_request(
+  cJSON *json_item = NULL;
+  log_info(json_logger_id, "[%s:%d] %s\n", __func__, __LINE__, obj);
+
+  if (json_obj == NULL) {
```
you can use `JSON_CHECK_ERROR` to validate JSON object and handle error response.

oopsmonk commented on cclient/serialization/json/get_balances.c:
```diff
@@ -56,6 +56,95 @@ retcode_t json_get_balances_serialize_request(
   return ret;
 }
 
+retcode_t json_get_balances_deserialize_request(const serializer_t *const s,
```
let's use `TYPE const * const IDENTIFIER` instead of `const TYPE *const IDENTIFIER` for keeping consistency and readability.

thibault-martinez replied:
I'll add that to `CONTRIBUTING.md` at some point 👍

howjmay committed:
use JSON_CHECK_ERROR and change arg order

howjmay committed:
fix miss one

oopsmonk commented on cclient/serialization/json/get_balances.c:
```diff
@@ -56,29 +56,92 @@ retcode_t json_get_balances_serialize_request(
+    goto end;
+  }
+
+  ret = json_array_to_hash243_queue(json_obj, "tips", &req->tips);
```
`tips` is an optional parameter, no need `goto` here.

oopsmonk replied:
This error code from `tips` element doesn't mean failed to `json_get_balances_deserialize_request`, `tips` is optional.

oopsmonk commented on cclient/serialization/json/get_balances.c:
```diff
@@ -56,29 +56,92 @@ retcode_t json_get_balances_serialize_request(
-    return RC_CCLIENT_JSON_PARSE;
+  JSON_CHECK_ERROR(json_obj, json_item, json_logger_id);
+
+  ret = json_get_uint8(json_obj, "threshold", &req->threshold);
```
Could you please help to change `threshold` as uint8_t and the comment in `cclient/request/get_balances.h`?
```
typedef struct {
  /**
   * The confirmation threshold between 0 and 100(inclusive).
   *  Should be set to 100 for getting balance by counting only confirmed transactions.
   */
  uint8_t threshold;
```

howjmay replied:
sure

howjmay committed:
`tips` not jump, add commet

since `tips` is optional parameter, so we serialize it whenever we need.

revise the commet of threshold in `cclient/request/get_balances.h`

howjmay committed:
tips deserialize err handling

oopsmonk commented on cclient/serialization/json/get_balances.c:
```diff
@@ -33,11 +33,90 @@ retcode_t json_get_balances_serialize_request(
+    cJSON_free((void *)json_text);
+  }
+
+  cJSON_Delete(json_root);
```
remove these two.

oopsmonk replied:
I mean remove
```
  cJSON_Delete(json_root);
  return ret;
```
b/c it's duplicate with the following
```
err:
  cJSON_Delete(json_root);
  return ret;
```

oopsmonk commented on cclient/serialization/json/get_balances.c:
```diff
@@ -33,11 +33,90 @@ retcode_t json_get_balances_serialize_request(
+  }
+  if (cJSON_GetObjectItemCaseSensitive(json_obj, "tips")) {
+    ret = json_array_to_hash243_queue(json_obj, "tips", &req->tips);
+    if (ret) {
```
remove if condition since here is the end.

oopsmonk commented on cclient/serialization/json/helpers.c:
```diff
@@ -59,6 +59,29 @@ retcode_t utarray_to_json_array(UT_array const* const ut,
+    uint64_t* p = NULL;
+    while ((p = (uint64_t*)utarray_next(ut, p))) {
+      char buffer[20];
+      sprintf(buffer, "%lld", *p);
```
should be `PRIu64` not `long long int`

oopsmonk commented on cclient/serialization/json/helpers.c:
```diff
@@ -59,6 +59,29 @@ retcode_t utarray_to_json_array(UT_array const* const ut,
+
+    uint64_t* p = NULL;
+    while ((p = (uint64_t*)utarray_next(ut, p))) {
+      char buffer[20];
```
move `buffer` out the while loop.

howjmay committed:
couple fixes

remove if condition

use PRIu64 instead of lld

move buffer out of while loop

th0br0 commented on cclient/serialization/json/get_balances.c:
```diff
@@ -33,22 +33,94 @@ retcode_t json_get_balances_serialize_request(
+    }
+  }
+
+  return ret;
```
remove, the label below will take care of this, also, this return will create a memleak

th0br0 commented on cclient/serialization/json/get_balances.c:
```diff
@@ -33,22 +33,94 @@ retcode_t json_get_balances_serialize_request(
+  }
+
+end:
+
```
remove empty line after `end:`

th0br0 commented on cclient/serialization/json/get_balances.c:
```diff
@@ -33,22 +33,94 @@ retcode_t json_get_balances_serialize_request(
+    goto end;
+  }
+  if (cJSON_GetObjectItemCaseSensitive(json_obj, "tips")) {
+    ret = json_array_to_hash243_queue(json_obj, "tips", &req->tips);
```
`json_array_to_hash243_queue` already checks for whether the item exists & will return an error code if it doesn't, so no need to wrap this in an `if`

th0br0 commented on cclient/serialization/json/get_balances.c:
```diff
@@ -33,22 +33,94 @@ retcode_t json_get_balances_serialize_request(
   }
 
-  cJSON_Delete(json_root);
   return ret;
```
remove this `return ret`

howjmay committed:
remove unnecessaey things

oopsmonk commented on cclient/serialization/json/get_balances.c:
```diff
@@ -33,23 +33,88 @@ retcode_t json_get_balances_serialize_request(
+  cJSON_AddItemToObject(json_root, "milestoneIndex",
+                        cJSON_CreateNumber(res->milestoneIndex));
+
+  json_text = cJSON_PrintUnformatted(json_root);
```
json_text need to be freed

oopsmonk commented on cclient/serialization/json/get_balances.c:
```diff
@@ -33,23 +33,88 @@ retcode_t json_get_balances_serialize_request(
     if (ret == RC_OK) {
       strncpy(out->data, json_text, len);
     }
-    cJSON_free((void *)json_text);
```
please bring it back, json_text need to be freed

oopsmonk reviewed (changes requested):
test your PR locally via `bazel test -c dbg --config asan //cclient/serialization/json/tests:get_balances`

howjmay committed:
Add if determine RC_CCLIENT_JSON_KEY

howjmay commented:
Hi @oopsmonk!
Is there anything that I need to modify?

oopsmonk commented:
> Hi @oopsmonk!
> Is there anything that I need to modify?

please rebase your PR, thanks.

howjmay committed:
Merge branch 'develop' into get_balance_serialize_res_deserialize_req

thibault-martinez committed:
Merge branch 'develop' into get_balance_serialize_res_deserialize_req

thibault-martinez committed:
Merge branch 'develop' into get_balance_serialize_res_deserialize_req
